Add Contributor's guide (#134)
UXARRAY/uxarray: docs/contributing.rst

.. _contributing:

.. toctree::
 :maxdepth: 2
 :numbered:

Contributor's Guide
===================

Welcome to the Contributor's Guide for UXarray!

> So, welcome to the team! If you are reading this document, we hope that
> you are already or soon-to-be a UXarray contributor! Please keep reading!
> 

Reviewer: ```suggestion
Welcome to the team! If you are reading this document, we hope that
you are already or soon-to-be a UXarray contributor, please keep reading!

```

1. Overview
-----------

UXarray is a community-owned, open-development effort that is a result
of the collaboration between `Project Raijin <https://raijin.ucar.edu/>`_
and DOE’s `SEATS Project <https://seatstandards.org/>`_ ]. Even though the

> Uxarray team has initiated and been expanding this package, outside

Reviewer: ```suggestion
UXarray team has initiated and been expanding this package, outside
```

contributions are welcome and vital for its long-term sustainability.
Therefore, we invite other community members to become part of this
collaboration at any level of contribution.

1.1. Why the name "UXarray"?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> We have created UXarray by composing `Xarray <https://docs.xarray.dev/en/stable/>`_,
> a Pangeo ecosystem package commonly-used  for structured grids
> recognition, to support reading and recognizing unstructured grid
> model outputs. We picked the name "UXarray" and preferred to
> capitalize the first two letters to emphasize it is Xarray for
> Unstructured grids.

Reviewer: ```suggestion
We have created UXarray by composing `Xarray <https://docs.xarray.dev/en/stable/>`_,
a Pangeo ecosystem package commonly-used for structured grids
recognition, to support reading and recognizing unstructured grid
model outputs. We picked the name "UXarray" and preferred to
capitalize the first two letters to emphasize it is Xarray for
Unstructured grids.
```

I'm not sure what you mean by "composing" here

Author: I (was trying to) refer to the OOP concept, composition, for Xarray dataset, but the word "composing" apparantly won't work for that. How about something like this:

"We have created UXarray based on `Xarray <https://docs.xarray.dev/en/stable/>`_ (via composition of a Xarray dataset object) ..."

Reviewer: Oh! Yes that is much more clear for me

1.2. Many ways to contribute
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

There are many different ways to contribute to UXarray. Anyone can, for
example,

> - Write or revise user documentation (including this document)

Reviewer: ```suggestion
- Write or revise documentation (including this document)
```

- Implement data analysis operators for unstructured grids from scratch
  or from their workflows

- Develop example notebooks that demonstrate how a particular function
  is used

- Answer a support question

- Simply request a feature or report a bug

All of these activities are signicant contributions to the on-going
development and maintenance of UXarray.

1.3. About this guide
^^^^^^^^^^^^^^^^^^^^^

The UXarray content is hosted on GitHub, and through this document,
we aim to ease the community's experience with contributing to this project.
However, this guide might still be missing case-specific details; please do
not hesitate to reach out to us to consult any such cases.

Lastly, much of the information in this guide has been co-opted from the
`GeoCAT <https://geocat.ucar.edu/pages/contributing.html>`_ project and
`Project Pythia <https://projectpythia.org/contributing.html>`_.

1.4. Project Specific Resources
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some important UXarray resources are as follows:

- `Uxarray GitHub repository <https://github.com/UXARRAY/uxarray>`_ houses
  the open-source code base along with some significant documentation such

>   as “Readme”, “Installation”, “Contributing”, and UXarray draft API.

Reviewer: Should we make these links?

Author: Thanks for pointing this out! I also noticed we should provide UXarray documentation as another resource here and link to Installation and Contributing links as well as some others from under that.

All of these activities are signicant contributions to the on-going
development and maintenance of UXarray.

1.3. About this guide
^^^^^^^^^^^^^^^^^^^^^

The UXarray content is hosted on GitHub, and through this document,
we aim to ease the community's experience with contributing to this project.
However, this guide might still be missing case-specific details; please do
not hesitate to reach out to us to consult any such cases.

Lastly, much of the information in this guide has been co-opted from the
`GeoCAT <https://geocat.ucar.edu/pages/contributing.html>`_ project and
`Project Pythia <https://projectpythia.org/contributing.html>`_.

1.4. Project Specific Resources
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some important UXarray resources are as follows:

- `Uxarray GitHub repository <https://github.com/UXARRAY/uxarray>`_ houses
  the open-source code base along with some significant documentation such
  as “Readme”, “Installation”, “Contributing”, and UXarray draft API.

- `UXarray draft API <https://github.com/UXARRAY/uxarray/blob/main/docs/user_api/uxarray_api.md>`_
  tentatively shows the eventual list of UXarray functionalities. It is
  open to community feedback and inputs!

>   - Please create a
>     GitHub `Discussion <https://github.com/UXARRAY/uxarray/discussions>`_ or
>     `Issue <https://github.com/UXARRAY/uxarray/issues>`_ if you feel there
>     should be change in any way in this document!

Reviewer: ```suggestion
  - Please create a
    GitHub `Discussion <https://github.com/UXARRAY/uxarray/discussions>`_ or
    `Issue <https://github.com/UXARRAY/uxarray/issues>`_ if you feel any changes
    should be made to this document!
```

  and :ref:`api`.

* `UXarray draft API <https://github.com/UXARRAY/uxarray/blob/main/docs/user_api/uxarray_api.md>`_
  tentatively shows the eventual list of UXarray functionalities. It is
  open to community feedback and inputs!

  - Please create a
    GitHub `Discussion <https://github.com/UXARRAY/uxarray/discussions>`_ or
    `Issue <https://github.com/UXARRAY/uxarray/issues>`_ if you feel any changes
    should be made to this document!

2. Configuring GitHub and Git, and Setting Up Python Environment
================================================================

In this section, we provide details about how to configure Github and Git
in order to work with the UXarray GitHub repository. This is because most
contributions, such as adding a new function or making changes to an
existing one, writing a usage example, or making some modifications to this
Contributor's Guide requires operating directly on the GitHub repository.

Contributing to a GitHub repository follows almost the same process by
any open development Python project maintained on GitHub.
However, it can still seem complex and somewhat varied from one project to

> another. As such, we will provide an overview of GitHub here and refer the

Reviewer 2: I think we should assume contributors have basic git and github knowledge.

Reviewer: I think this might be a reasonable assumption, but I've definitely had some users or potential contributors to other projects not have a github. We should keep this section in, I think, to err on the side of not assuming prior knowledge as much as possible.

Reviewer 2: I do.  I really want developers to have already learned the basics of git/github on their own before contributing to uxarray.  The presences of this material implies we want total novices contributing and I don't think we do.

Reviewer: @rljacob I can see where you're coming from, but a total novice to git/github might not be a novice with respect to python coding, algorithm development, or good ideas.

Reviewer 2: I don't want to discourage git novices but I don't want UXarray to take on the responsibility for teaching them git/github.  We should just point them to our favorite websites for that and then focus on the UXarray-specific parts of contributing code.

Author: Ok, this makes sense! I have removed the details of GitHub/Git configuration and instead put some text that refers the refer to the comprehensive guides of those. Please let me know if the new look suffices for your concerns.

.. _contributing:

*******************
Contributor's Guide
*******************

Welcome to the Contributor's Guide for UXarray!

Welcome to the team! If you are reading this document, we hope that
you are already or soon-to-be a UXarray contributor, please keep reading!

1. Overview
===========

UXarray is a community-owned, open-development effort that is a result
of the collaboration between `Project Raijin <https://raijin.ucar.edu/>`_

> and DOE’s `SEATS Project <https://seatstandards.org/>`_ ]. Even though the

Author: Done.

UXarray team has initiated and been expanding this package, outside
contributions are welcome and vital for its long-term sustainability.
Therefore, we invite other community members to become part of this
collaboration at any level of contribution.

1.1. Why the name "UXarray"?
----------------------------

We have created UXarray based on `Xarray <https://docs.xarray.dev/en/stable/>`_
(via composition of a Xarray dataset object), a Pangeo ecosystem package
commonly-used  for structured grids recognition, to support reading and
recognizing unstructured grid model outputs. We picked the name "UXarray"
and preferred to capitalize the first two letters to emphasize it is Xarray
for Unstructured grids.

1.2. Many ways to contribute
----------------------------

There are many different ways to contribute to UXarray. Anyone can, for
example,

* Write or revise documentation (including this document)
* Implement data analysis operators for unstructured grids from scratch
  or from their workflows
* Develop example notebooks that demonstrate how a particular function
  is used
* Answer a support question

> * Simply request a feature or report a bug

Author: Done!

language. Because there are so many Python packages available,
maintaining interoperability between them is a huge challenge. To overcome
some of these difficulties, we strongly recommend the use of `Anaconda
<https://www.anaconda.com/download/>`_ or `Miniconda
<https://conda.io/miniconda.html/>`_ as a package manager to manage your
Python ecosystem. These package managers allow you to create a separate,
custom Python environment for each specific Python set of tools. Yes, this
unfortunately results in multiple copies of Python on your system, but it
greatly reduces breaking toolchains whenever a change is made to your Python
environment (and is more reliable than any other solution we’ve encountered).
Also, the use of Anaconda/Miniconda is standard practice for working with
Python in general, not simply for using UXarray.

To configure your Python environment:

1. Install either `Anaconda <https://www.anaconda.com/download/>`_ or
`Miniconda <https://conda.io/miniconda.html/>`_.

   * We recommend Miniconda as it does not install unnecessarily many
     packages that will not be needed for UXarray development.

> 

Author: Done! Please let me know if the new version suffices.

> 

Reviewer: Looks good!

2. Make sure your conda is up to date by running this command from the
terminal::

    $ conda update conda

At this point, you will have a current version of conda available on your
system. Before using your Python environment to work on UXarray development,
you’ll need to perform additional steps that are going to be described in
the next section.

3. How UXarray Uses Git/GitHub
==============================

UXarray uses the `GitHub Flow
<https://docs.github.com/en/get-started/quickstart/github-flow>`_ model
for its workflow. UXArray also uses an automated formatter on all commits
in local development environment in order to ensure code formatting.
This changes the normal workflow slightly, so in order to avoid any

> confusions, follow these steps:

Author: There is indeed a separate section for pre-commits configuration and usage, and point the reader to that, I have now added the following to this intro text:

"UXArray also uses an automated formatter, which is described in `3.4. Install and Setup Pre-commit Hooks`_ in detail, on all commits in local development environment in order to ensure code formatting." 

Would this suffice for what you were thinking?

> confusions, follow these steps:

Reviewer: Haha yes, I think I mentioned that I found the section farther down. I think with the linking it's good!

3.1. Select an issue to work on
-------------------------------

Virtually any work should be addressing an issue. There are a few options to
select an issue to work on:

1. First, check the existing `UXarray issues
<https://github.com/UXARRAY/uxarray/issues>`_. These issues might have been
created from either the

* `UXarray draft API
  <https://github.com/UXARRAY/uxarray/blob/main/docs/user_api/uxarray_api.md>`_
  entries,
* An entry in a `UXarray Discussion <https://github.com/UXARRAY/uxarray/discussions>`_
* A feature request, bug report, or any finding of the developers or users
  as a work to-do.

2. If you are going to work on one of those issues above, self-assign that
issue before you start working.

3. Or else, if you are going to work on something else, create an issue. The
title, description, and label of an issue could be very helpful for the team to

> triage the issue and move forward with the work needed.

Author: Done.

> 

Author: Line break helps the bullet points be generated properly.

* Title needs to reflect the purpose of the issue. Some examples:
  - Start with "Add ..." if it is a new functionality needed

>   - Start with "Error ...", "Exception ...", or "Bug ..." if that is a bug

Author: Thanks for catching! Just fixed!

    request (PR) is opened, it will require others' thorough review and discussion
    before your changes can take place in the project. Hence, a rule of thumb about
    pull requests should be that they should target logically connected, as atomic
    as possible work.

.. note::
    If you need any clarification/discussion about any requirements, or if you
    think the implementation of that issue will require significant changes to
    the code, design, documentation, etc. that issue itself is the right place
    to manage such discussions with the other UXarray'ers. Don't hesitate to
    ask ad-hoc meetings, etc.

    Do not forget, early requirements analysis, specifications, and design
    discussions can avoid redundat code review and modifications later!

3.2. Fork or locally clone the UXarray repository
-------------------------------------------------

Let us first make a decision of whether to fork or locally clone:

3.2.1. Should you fork?
^^^^^^^^^^^^^^^^^^^^^^^

"Fork"ing a repository as described below in this section creates a copy of
the repository under your own account or any GitHub organization of

> which you are a member on GitHub. Forking can be advisable for cases such as:

Author: Agreed!

* You want to safely make changes to the forked repository contents without changing
  the actual UXarray repository. This is because any changes you make to your
  fork will only be seen by you until you are ready to share them with others,
  and hopefully “merge” your changes into UXarray.
* You are not a regular contributor to the UXarray repository; thus, from time
  to time, you would sync your fork to UXarray and make one or more contributions.
* You are planning to initiate a new project by altering the forked repo
  significantly from UXarray, but this case is out of this guide's scope.

You can refer to the `Atlassian's Forking workflow
<https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow>`_
tutorial to further learn about forking.

If you decide on forking the UXarray repository, here is how to do that:

> **Forking the UXarray repository**

Author: I thought as it was already under the section header "3.2.1. Should You Fork?", we should be good this way.

Refer to GitHub's `Forking a repository
<https://docs.github.com/en/get-started/quickstart/fork-a-repo>`_ guide and apply
instructions regarding the UXarray repository.

.. note::
    The above GitHub guide about forking a repository will also walk you through
    cloning your forked repository into your local work environment. Please do not
    forget to follow those instructions as well. Or, refer to `GitHub's Cloning a
    repository <https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository>`_
    guide, and apply those instructions for the UXarray fork that is listed under
    your GitHub account or organization.

After these steps, you will have two copies of the forked UXarray repo, one remote
and one local.

3.2.2. Should you locally clone instead?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In contrast to above cases that might be better suitable for fork, a regular
UXarray contributor who is comfortable with working on their local clone
of the actual UXarray repository and making their changes immediately viewable by
the other UXarray'ers (i.e. after "push"ing their "commit"s) can choose to locally
clone the UXarray repository.

> **Locally cloning the UXarray repository**

Author: Similar to the previous one.

Refer to `GitHub's Cloning a repository
<https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository>`_
guide and apply those instructions for the actual `UXarray repository
<https://github.com/UXARRAY/uxarray>`_.

.. note::
    Regardless of you either fork or clone, there will be a local directory created
    in the name "uxarray" (unless you specified a different name at the step with the

>     :code:`git clone` command). You can type the following command in the terminal/shell to

Reviewer: ```suggestion
    ``git clone`` command). You can type the following command in the terminal/shell to
```

Author: Done!

    go into your local UXarray repository::

>         $ cd uxarray

Author: Yes, most likely, but it may still exist just as a hint.

3.3. Configure UXarray Conda Environment
----------------------------------------

In your local UXarray directory, creating a UXarray conda environment is needed for
development purposes. Use the following commands for this::

    $ conda env create --file ci/environment.yml

>     $ conda activate uxarray_build

Author: I couldn't find the previous suggestion with mamba, but I think conda should still be here in this guide

THe above commands will use the :code:`environment.yml` conda environment definition
file that is hosted under the :code:`ci` folder and create a conda environment with
the name "uxarray_build". Once you activate that environment with the help of the
second command, you will be able to develop UXarray codes in your local configuration.

> 3.4. Install and Setup Pre-commit Hooks

Author: In-text links have been added now

`.pre-commit-config.yaml <https://github.com/UXARRAY/uxarray/blob/main/.pre-commit-config.yaml>`_
file, so there is no action needed), and install pre-commit as described below. Pre-commit
manages the installation and execution of the hooks specified.

Hooks are run on every commit to automatically point out issues in code formatting such
as the number of characters in each line, missing semicolons, trailing whitespace, debug
statements, etc. By pointing these issues out before code review, this allows a code
reviewer to focus on the architecture of a change while not wasting time with trivial
style nitpicks. Refer to the `Pre-commit documentation <https://pre-commit.com/#intro>`_
for further learning.

.. note::
    We also use pre-commit GitHub Actions workflow to make sure every code contribution
    to our UXarray's GitHub repository (i.e. pull request) aligns with our code standards.
    Therefore, the code changes that are not being checked through local pre-commit hooks
    will eventually be tested against our workflow in the GitHub server as described in
    `TODO`_. If there are any issues with the code format, it will lead the pull request

>     to fail pre-commit checks.

Author: It was a personal refernce to address before this gets merged. Addressed already. Thanks though!

3.4.1. Pre-commit setup
^^^^^^^^^^^^^^^^^^^^^^^

If you configured your Conda environment via the instructions in `3.3. Configure UXarray

> Conda Environment`_, you will have the :code:`pre-commit` package already installed in your

Author: Done

> environment (Otherwise, you will need to run :code:`conda install -c conda-forge pre_commit`

Author: Decided to keep using `:code:` format for only actual code referrals in-text. Otherwise applied the suggestion.

to get it installed into your Conda environment for UXarray development). The only thing you
will have to do additionally in order to set up pre-commit is to run the following command
in your terminal in the UXarray root directory::

    $ pre-commit install

At this step, you are good to go with your pre-commit hooks to check your future commits.
If, at any time, you'd like to run pre-commit hooks on your all files, you can run the
following command::

    $ pre-commit run --all-files

3.5. Use Feature Branches
-------------------------

> In your local clone, make a new branch off of the :code:`main` branch (This is way to go

Author: Done

most of the time, but there might be specific cases where, for example, a branch is needed
to be created off of another feature branch). Naming this branch, whenever applicable,
like the following is not required but may be much helpful for tracking purposes:
"issue_XXX" where XXX is the number of the issue or something that is representative of
the changes you’re making, e.g. "do-this-work", "add-that-function", etc.

> Here are example commands that assume, you are checking out the :code:`main` branch first,

Author: Done

pulling  from the remote server to have everything in your local up-to-date, and creating a new

> branch off of :code:`main`::

Author: Done

    $ git checkout main
    $ git pull
    $ git checkout -b <new_branch>

Once you create the new branch, you are good to go with your local changes in the UXarray
directory!

3.7. Local changes, commits, and pushes
---------------------------------------

The local development process can very basically be itemized as follows:

> 1. Make change(s) to your local copy of the UXarray repository

Author: Done

   .. note::
        Please refer to `3.7.2. Common Elements of Most Pull Requests`_ to make sure your local
        changes have all of the elements they should cover.

2. Changes you’ve made locally must be “committed” (merged) to your local repository

>    (the .git subdirectory) using Git, but before the commit, you need to add them to the

Author: Removed the text, so no need for this

>    "staged" changes for commit.

Author: Rephrased.

   * You can see any uncommitted changes you’ve made to your local copy of the repository by
     running the following command from anywhere (any directory) within the directory where

>      you ran :code:`git checkout`::

Author: Reject

        $ git status

   * So, add changed file(s) into the staged changes to be included in a single commit::

        $ git add PATH_TO_NEW_FILE

>      where `PATH_TO_NEW_FILE` is the path name of the newly created file.

Reviewer 3: ```suggestion
     where ``PATH/TO/NEW/FILE`` is the path name of the newly created file.
```
For this, it may also be better to use slashes instead to show the correct format for a path

>      where `PATH_TO_NEW_FILE` is the path name of the newly created file.

Author: Both done

3. Now, commit the staged change(s)::

        $ git commit -m "Descriptive comment about what this commit does"

   * Limiting the commit to file(s) changed for an atomic task would be very much
     helpful in cases you need to review and maybe revert commits.

   * Do not forget that `pre-commit hooks` will check your code changes at this
     point. If some of those hooks fail, you will need to add those files again
     and attempt to make the same commit again as the hooks will have changed those
     file(s) to make them comply with the code formatting rules.

   * A good practice is to run::

        $ git status

     after your commit to verify everything looks as expected.

4. Push your commit(s) into the remote repository::

    $ git push

   or::

    $ git push --set-upstream origin <your-branch-name>

>    if you haven't created a corresponding remote branch for it yet).

Author: Done, thanks!

.. note:: Remember that we are still not proposing to merge our work into the UXarray
    GitHub repository, which will be described in the next subsection, `3.7. Pull
    Requests`_.

3.7. Pull Requests
------------------

> Pull requests (PR) are GitHub's way for developers to propose and collaborate on changes

Author: Thanks for catching!

to a repository. If you are interested in learning the foundations about pull requests,
please refer to GitHub's `Pull requests <https://docs.github.com/en/pull-requests>`_
documentation.

Once you have completed making changes to your local copy of the UXarray repository,
pushed them all, and are ready to have your changes merged into the repository on GitHub,
you need to submit a PR asking the UXarray maintainers to consider your merge request.

The merge will occur between your personal branch that should have all of your commits
from local in the GitHub repository (either in your fork or in the actual UXarray repo)

> and the :code:`main` branch, if not other, in the UXarray's GitHub repository. There

Reviewer: ```suggestion
and the ``main`` branch, if not other, in the UXarray's GitHub repository. There
```

    as well as some guidance from others. Please read this GitHub `blog
    <https://github.blog/2019-02-14-introducing-draft-pull-requests/>`_ about draft PRs.

Please refer to Github's `Creating a pull request
<https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request>`_
guide for the instructions. We are especially avoiding to detail these instructions
here as there are manly referrals to GitHub's graphical user interface, which might be
changed in the future.

.. note::
    Below are significant things about pull requests that can be very helpful throughout
    the entire contribution process (i.e. review and merge) process when performed:

3.7.1. Review your changes
^^^^^^^^^^^^^^^^^^^^^^^^^^

Before you finalize opening the actual PR, it is a good practice to review the changes
that you’ve made. You should be able to review all of the changes that will go into
this PR just before you press the `Create pull request` button.

If there is any changes you want to make in the PR, you can delay creating the PR
and push new commits, revert existing changes, etc. You can then create the PR.

> 3.7.2. Common Elements of Most Pull Requests

Reviewer: ```suggestion
3.7.2. Common Elements of Pull Requests
```

Author: Done

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Despite the fact that pull requests can differ regarding their purposes (e.g.
correction of a single typo in only one file could make a PR), most of the PRs may
consist of code changes that should, most of the time, include some other elements
with it. Having all of such elements addressed in a PR could make the review and
merge process a lot easier. Assuming such code changes, these are what might
accompany them:

3.7.2.1. Unit tests
~~~~~~~~~~~~~~~~~~~

Virtually all new Uxarray code needs to include unit tests of it implemented.
The UXarray project makes use of diverse technologies for unit testing as
follows:

* All the unit tests of every single Python module (i.e. `.py` file) should be

>   implemented as a separate test script under the :code:`\test` folder of the

Reviewer: ```suggestion
  implemented as a separate test script under the ``test/`` folder of the
```

Author: Done

  repository's root directory.

* The `pytest <https://docs.pytest.org/en/stable/contents.html>`_ testing framework
  is used as `runner` for the tests. If you configured your Conda environment via
  the instructions in `3.3. Configure UXarray Conda Environment`_, you will have

>   the :code:`pytest` package already installed in your environment (Otherwise, you

Reviewer: ```suggestion
  the ``pytest`` package already installed in your environment (Otherwise, you
```

You might also want to check if this is attempting to link to pytest docs, if that's what you're going for.

Author: One link in the beginning suffices I think

>   will need to run :code:`conda install -c conda-forge pytest` to get it installed

Reviewer: ```suggestion
  will need to run ``conda install -c conda-forge pytest`` to get it installed
```

Author: Reject as this is an in-text reference to actual code (It can help review the rst file in the future)

  into your Conda environment for UXarray development).

> * Test scripts themselves are not intended to use `pytest` through implementation.

Reviewer: ```suggestion
* Test scripts themselves are not intended to use ``pytest`` through implementation.
```

Author: Done

> Instead, `pytest` should be used only for running test scripts as follows::

Reviewer: ```suggestion
Instead, ``pytest`` should be used only for running test scripts as follows::
```

Author: Done

    $ pytest test/<test_script_name>.py

> Not using `pytest` for implementation allows the unit tests to be also run

Reviewer: ```suggestion
Not using ``pytest`` for implementation allows the unit tests to be also run
```

Reviewer 4: Looks good, we can list all the different ways to run, it doesn't have to be pytest, I run it as 
`
python -m unittest test.test_grid.TestGrid`

Author: Thanks; done!

by using::

    $ python -m unittest test/<test_script_name>.py

Also, all of the test scripts can be run at once with the following command::

    $ pytest test

> * Python's unit testing framework, `unittest
> <https://docs.python.org/3/library/unittest.html>`_ is used for implementation of

Reviewer: ```suggestion
* Python's unit testing framework, `unittest
<https://docs.python.org/3/library/unittest.html>_` is used for implementation of
```

Author: Reject. The correct rst way of URLs is how it is used here.

the test scripts.

> * Reference results (i.e. expected output or ground truth for not all but the most cases)
> need not to be magic values (i.e. they need to be justified and/or documented).

Reviewer: ```suggestion
* Reference results (i.e. expected output or ground truth for not all but the most cases)
should not be magic values (i.e. they need to be justified and/or documented).
```

Author: Done

* Recommended, but not mandatory, implementation approach is as follows:

  - Common data structures, variables and functions,  as well as
    expected outputs, which could be used by multiple test methods throughout
    the test script, are defined either under a base test class or in the very
    beginning of the test script for being used by multiple unit test cases.

>   - Any group of testing functions dedicated to testing a particular
>     phenomenon (e.g. a specific edge case, data structure, etc.) is
>     implemented by a class, which inherits `TestCase` from Python's
>     `unittest` and likely the base test class implemented for the purpose
>     mentioned above.

Reviewer: ```suggestion
  - Any group of testing functions dedicated to testing a particular
    phenomenon (e.g. a specific edge case, data structure, etc.) is
    implemented by a class, which inherits ``TestCase`` from Python's
    ``unittest`` and likely the base test class implemented for the purpose
    mentioned above.
```

Author: Done

  - Assertions are used for testing various cases such as array comparison.

  - Please see previously implemented test cases for reference of the
    recommended testing approach

> .. note:: Our test suite that includes all the unit tests is executed automarically
> for PRs with the help of GitHub Actions workflows to ensure new code passes tests.
> Hence, please check `TODO`_ to make sure your PR tests are all passing before asking
> others to review your work.

Reviewer: ```suggestion
.. note:: 
Our test suite that includes all the unit tests is executed automatically
for PRs with the help of GitHub Actions workflows to ensure new code passes tests.
Hence, please check `TODO` to make sure your PR tests are all passing before asking
others to review your work.
```
![image](https://user-images.githubusercontent.com/38434768/191373430-c7e2e19f-f463-47c4-8acc-9de42de085ac.png)

Author: Fixed; thanks!

3.7.2.2. Docstrings
~~~~~~~~~~~~~~~~~~~

> All Python functions must contain a `Google Style Python
> <https://github.com/google/styleguide/blob/gh-pages/pyguide.md>`_ *docstring* (i.e.
> triple quoted comment blocks). These docstrings are accessed from the Python interpreter
> whenever the user types::

Reviewer: ```suggestion
All Python functions must contain a `Google Style Python
<https://github.com/google/styleguide/blob/gh-pages/pyguide.md>`_ docstring (i.e.
triple quoted comment blocks). These docstrings are accessed from the Python interpreter
whenever the user types::
```

    help <FUNCTION_NAME>

They are also automatically converted into web-accessible documentation available from
the `UXarray documentation <https://uxarray.readthedocs.io/>`_.

The docstrings must contain:

1. A brief description of the functionality provided. What does this function do?

2. If available, references to the algorithm or implementation employed

3. A complete description of arguments and return values in Google format

   * Please refer to the existing functions that already have this

4. One or more very short usage examples that demonstrates how to invoke
   the function, and possibly what to expect it to return

   * No need for these examples to actually be executable

   * If a usage example is longer than a handful of lines, a more complete

>      example may be created instead, referring to `TODO `_.

Reviewer: Same as similar above comments

Author: Fixed

3.7.2.3. Documentation
~~~~~~~~~~~~~~~~~~~~~~

As we mentioned a few times throughout this guide, Uxarray has a static `documentation
<https://uxarray.readthedocs.io/>`_ :ref:`index` page that is being generated automatically from the
repository's code structure. However, there needs to be some manual additions to the
proper documentation index file(s) for the automation to work. The following index files
are used for UXarray documentation (directories relative from the root):

    `docs/user_api/index.rst`

>     `docs/internal_api/index.rst`

Reviewer: I think this bit also looks a bit funny when generated
![image](https://user-images.githubusercontent.com/38434768/191373808-28a13e70-1b2f-4d37-8b5b-4b7f731d41db.png)

Author: Fixed

> which allows the following documentation to be autoamtically generated, respectively:

Reviewer: ```suggestion
which allows the following documentation to be automatically generated, respectively:
```

Author: Fixed

    :ref:`user_api/index`
    :ref:`internal_api/index`

> 

Reviewer: Could be a good use for cards!

Author: Thanks for the suggestion! I had changed the structure of that part into a more paragraph way in the meantime. Please let me know if you still think cards might be applicable for that part.

That being said, the code changes, which might be a new function implementation or some

> modifications to existing ones, must be added to the appropriate `index.rst`

Reviewer: ```suggestion
modifications to existing ones, must be added to the appropriate ``index.rst``
```

this PR just before you press the `Create pull request` button.

If there is any changes you want to make in the PR, you can delay creating the PR
and push new commits, revert existing changes, etc. You can then create the PR.

3.7.3. Common Elements of Most Pull Requests
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Despite the fact that pull requests can differ regarding their purposes (e.g.
correction of a single typo in only one file could make a PR), most of the PRs may
consist of code changes that should, most of the time, include some other elements
with it. Having all of such elements addressed in a PR could make the review and
merge process a lot easier. Assuming such code changes, these are what might
accompany them:

3.7.3.1. Unit tests
~~~~~~~~~~~~~~~~~~~

Virtually all new Uxarray code needs to include unit tests of it implemented.
The UXarray project makes use of diverse technologies for unit testing as
follows:

* All the unit tests of every single Python module (i.e. `.py` file) should be
  implemented as a separate test script under the :code:`\test` folder of the
  repository's root directory.

> * The `pytest <https://docs.pytest.org/en/stable/contents.html>`_ testing framework

Reviewer: This might be ignorance on my part, but why, if we're writing tests with unittest, do we run them with pytest? Why don't we just run them with unittest?

Also, depending on the answer to the above question, this might be a good place to mention the [practical differences](https://docs.pytest.org/en/7.1.x/how-to/usage.html#other-ways-of-calling-pytest) between running `pytest` directly vs `python3 -m pytest`

Author: A few benefits coming from using `pytest` are listed [here](https://docs.pytest.org/en/7.1.x/how-to/unittest.html#how-to-use-unittest-based-tests-with-pytest). I linked to that document in the part of the guide where we discuss `unittest` can be used instead of `pytest` as follows:

"a number of benefits/conveniences coming from using ``pytest`` can be
seen `<URL>` though"

I added a new command (`python -m pytest test/<test_script_name>.py`) and mentioned the benefit of using `python -m ...` instead of `pytest` as follows:

"the latter of which will also add the current directory to :code:`sys.path`"

terminal::

    $ conda update conda

.. warning::
    Don’t type the $ character. This simply indicates the command line prompt.

At this point, you will have a current version of conda available on your
system. Before using your Python environment to work on UXarray development,
you’ll need to perform additional steps that are going to be described in
the next section.

3. How UXarray Uses Git/GitHub
==============================

UXarray uses the `GitHub Flow
<https://docs.github.com/en/get-started/quickstart/github-flow>`_ model
for its workflow. UXarray also uses an automated formatter, which is
described in `3.4. Install and Setup Pre-commit Hooks`_ in detail, on
all commits in local development environment in order to ensure code
formatting. This changes the normal workflow slightly, so in order to
avoid any confusions, follow these steps:

3.1. Select An Issue to Work on
-------------------------------

Virtually any work should be addressing an issue. There are a few options to
select an issue to work on:

1. First, check the existing `UXarray issues
<https://github.com/UXARRAY/uxarray/issues>`_. These issues might have been
created from either the

> * `UXarray draft API

Reviewer 5: The markdown format seems to be off here

Author: Done; thanks for catching this!

relative from the root directory) are used for UXarray documentation to allow the
`User API <user_api/index.rst>`_ and `Internal API <internal_api/index.rst>`_,
respectively, to be automatically generated.

That being said, the code changes, which might be a new function implementation or some
modifications to existing ones, must be added to the appropriate ``index.rst``
file so that its documentation page is automatically generated.

.. caution::
    Please also ensure that you are changing the `UXarray draft API
    <https://github.com/UXARRAY/uxarray/blob/main/docs/user_api/uxarray_api.md>`_
    accordingly if you are proposing changes to the API (e.g. new functions/attributes,
    modification(s) to existing functions, etc.)

3.7.3.5. Usage examples
~~~~~~~~~~~~~~~~~~~~~~~

.. note::
    This may not be required for every PR and can be handled as separate PRs when
    needed. However, it would be a great practice to provide usage examples in the
    same PR, especially for demonstrating the use of complex UXarray functions.

The UXarray documentation houses `examples/<example-name>.ipynb` files (paths relative

> from the root directory) to provide `examples.rst>`_ to be automatically generated. If

Reviewer 5: ```suggestion
from the root directory) to provide `examples.rst`_ to be automatically generated. If
```

Author: Great catch; done!

* Implement data analysis operators for unstructured grids from scratch
  or from their workflows
* Develop example notebooks that demonstrate how a particular function
  is used
* Answer a support question
* Simply request a feature or report a bug

All of these activities are signicant contributions to the on-going
development and maintenance of UXarray.

1.3. About this guide
---------------------

The UXarray content is hosted on GitHub, and through this document,
we aim to ease the community's experience with contributing to this project.
However, this guide might still be missing case-specific details; please do
not hesitate to reach out to us to consult any such cases.

.. note::
    Much of the information in this guide has been co-opted from the
    `GeoCAT <https://geocat.ucar.edu/pages/contributing.html>`_ project and
    `Project Pythia <https://projectpythia.org/contributing.html>`_.

1.4. Project Specific Resources
-------------------------------

Some important UXarray resources are as follows:

* `UXarray GitHub repository <https://github.com/UXARRAY/uxarray>`_ houses
  the open-source code base along with some significant documentation such

>   as `Readme <https://github.com/UXARRAY/uxarray/blob/main/README.md>`_ and

Author: Done!

create a Python environment along with a package manager. When we use the
term “Python environment” here, we are referring to the Python programming
language. Because there are so many Python packages available,
maintaining interoperability between them is a huge challenge. To overcome
some of these difficulties, we strongly recommend the use of `Anaconda
<https://www.anaconda.com/download/>`_ or `Miniconda
<https://conda.io/miniconda.html/>`_ as a package manager to manage your
Python ecosystem. These package managers allow you to create a separate,
custom Python environment for each specific Python set of tools. Yes, this
unfortunately results in multiple copies of Python on your system, but it
greatly reduces breaking toolchains whenever a change is made to your Python
environment (and is more reliable than any other solution we’ve encountered).
Also, the use of Anaconda/Miniconda is standard practice for working with
Python in general, not simply for using UXarray.

To configure your Python environment:

1. Install either `Anaconda <https://www.anaconda.com/download/>`_ or
`Miniconda <https://conda.io/miniconda.html/>`_.

>    * We recommend Miniconda as it does not install unnecessarily many

Author: Done!

* Title needs to reflect the purpose of the issue. Some examples:
  - Start with "Add ..." if it is a new functionality needed
  - Start with "Error ...", "Exception ...", or "Bug ..." if that is a bug
* Use issue labels accordingly, e.g. "bug", "support", "enhancement",
  "documentation", etc.
* Clearly describe the issue (e.g. what the work should be, which parts of
  UXarray it would need to change, if known, etc.)

.. note::
    The work that addresses an issue will, most of the time, be eventually
    turned into a "pull request" (or maybe multiple in some cases) that lets you
    tell others about changes you're aiming to make on the project. Once a pull
    request (PR) is opened, it will require others' thorough review and discussion
    before your changes can take place in the project. Hence, a rule of thumb about
    pull requests should be that they should target logically connected, as atomic
    as possible work.

.. note::
    If you need any clarification/discussion about any requirements, or if you
    think the implementation of that issue will require significant changes to
    the code, design, documentation, etc. that issue itself is the right place

>     to manage such discussions with the other UXarray'ers. Don't hesitate to

Author: Done

    forget to follow those instructions as well. Or, refer to `GitHub's Cloning a
    repository <https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository>`_
    guide, and apply those instructions for the UXarray fork that is listed under
    your GitHub account or organization.

After these steps, you will have two copies of the forked UXarray repo, one remote
and one local.

3.2.2. Should you locally clone instead?
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

In contrast to above cases that might be better suitable for fork, a regular
UXarray contributor who is comfortable with working on their local clone
of the actual UXarray repository and making their changes immediately viewable by
the other UXarray'ers (i.e. after "push"ing their "commit"s) can choose to locally
clone the UXarray repository.

**Locally cloning the UXarray repository**

Refer to `GitHub's Cloning a repository
<https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository>`_
guide and apply those instructions for the actual `UXarray repository
<https://github.com/UXARRAY/uxarray>`_.

.. note::

>     Regardless of you either fork or clone, there will be a local directory created

Author: Thanks!

    Therefore, the code changes that are not being checked through local `pre-commit
    hooks` will eventually be tested against our workflow in the GitHub server as
    described in `3.7.4.2. GitHub Actions checks`_. If there are any issues with the
    code format, it will lead the pull request to fail the pre-commit checks.

3.4.1. Pre-commit setup
^^^^^^^^^^^^^^^^^^^^^^^

If you configured your Conda environment via the instructions in `3.3. Configure UXarray
Conda Environment`_, you will have the :code:`pre-commit` package already installed in your
environment (Otherwise, you will need to run :code:`conda install -c conda-forge pre_commit`
to get it installed into your Conda environment for UXarray development). The only thing you
will have to do additionally in order to set up pre-commit is to run the following command
in your terminal in the UXarray root directory::

    $ pre-commit install

At this step, you are good to go with your pre-commit hooks to check your future commits.
If, at any time, you'd like to run pre-commit hooks on your all files, you can run the
following command::

    $ pre-commit run --all-files

3.5. Use Feature Branches
-------------------------

> In your local clone, make a new branch off of the :code:`main` branch (This is way to go

Author: Done

most of the time, but there might be specific cases where, for example, a branch is needed
to be created off of another feature branch). Naming this branch, whenever applicable,

> like the following is not required but may be much helpful for tracking purposes:

Author: Done

from local in the GitHub repository (either in your fork or in the actual UXarray repo)
and the :code:`main` branch, if not other, in the UXarray's GitHub repository. There
might be some exceptions to this generic case, which can always be learned through
reading or discussed with the maintainers and community.

.. note::
    While pull requests are supposed to be based on changes that are ready to be tested,
    reviewed, and eventually be merged to repositories; there is an exception to this:
    Draft Pull Requests. We encourage Draft PRs if you just want to just start a
    conversation about your code that isn’t in any state to be judged, and get feedback
    as well as some guidance from others. Please read this GitHub `blog
    <https://github.blog/2019-02-14-introducing-draft-pull-requests/>`_ about draft PRs.

Please refer to Github's `Creating a pull request
<https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request>`_
guide for the instructions. We are especially avoiding to detail these instructions

> here as there are manly referrals to GitHub's graphical user interface, which might be

Reviewer 3: ```suggestion
here as they are mainly referrals to GitHub's graphical user interface, which might be
```
Depending on the intent of this sentence, this could also be changed to "there are many referrals..." instead.

Author: Done

changed in the future.

.. note::
    Below are significant things about pull requests that can be very helpful throughout
    the entire contribution process (i.e. review and merge) process when performed:

3.7.1. Link PR to Issue
^^^^^^^^^^^^^^^^^^^^^^^

Don't forget to link your PR to the issue you have been working on! This will help not
only keep track of issues easier but also get them closed automatically once your PR is
merged. You can refer to GitHub's `Linking a pull request to an issue
<https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue>`_
for this.

3.7.2. Review your changes
^^^^^^^^^^^^^^^^^^^^^^^^^^

Before you finalize opening the actual PR, it is a good practice to review the changes
that you’ve made. You should be able to review all of the changes that will go into
this PR just before you press the `Create pull request` button.

> If there is any changes you want to make in the PR, you can delay creating the PR

Reviewer 3: ```suggestion
If there are any changes you want to make in the PR, you can delay creating the PR
```

Author: Done!

and push new commits, revert existing changes, etc. You can then create the PR.

3.7.3. Common Elements of Most Pull Requests
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Despite the fact that pull requests can differ regarding their purposes (e.g.
correction of a single typo in only one file could make a PR), most of the PRs may
consist of code changes that should, most of the time, include some other elements
with it. Having all of such elements addressed in a PR could make the review and
merge process a lot easier. Assuming such code changes, these are what might
accompany them:

> 3.7.3.1. Unit tests

Reviewer 3: ```suggestion
3.7.3.1. Unit Tests
```

~~~~~~~~~~~~~~~~~~~

> Virtually all new Uxarray code needs to include unit tests of it implemented.

Reviewer 3: ```suggestion
Virtually all new UXarray code needs to include unit tests of the functionality implemented.
```
This sentence was a little clunky, this is a potential way to word it instead

Author: Done , thanks!

3.7.3.2. Docstrings
~~~~~~~~~~~~~~~~~~~

All Python functions must contain a `Google Style Python
<https://github.com/google/styleguide/blob/gh-pages/pyguide.md>`_ *docstring* (i.e.
triple quoted comment blocks). These docstrings are accessed from the Python interpreter
whenever the user types::

    help <FUNCTION_NAME>

They are also automatically converted into web-accessible documentation available from
the `UXarray documentation <https://uxarray.readthedocs.io/>`_.

The docstrings must contain:

1. A brief description of the functionality provided. What does this function do?

2. If available, references to the algorithm or implementation employed

3. A complete description of arguments and return values in Google format

   * Please refer to the existing functions that already have this

4. One or more very short usage examples that demonstrates how to invoke
   the function, and possibly what to expect it to return

   * No need for these examples to actually be executable

   * If a usage example is longer than a handful of lines, a more complete
     example may be created instead, referring to `3.7.3.4. Usage Examples`_.

3.7.3.3. Documentation
~~~~~~~~~~~~~~~~~~~~~~

> As we mentioned a few times throughout this guide, Uxarray has a static `documentation

Reviewer 3: ```suggestion
As we mentioned a few times throughout this guide, UXarray has a static `documentation
```

Author: Done!

3.7.4. After You Open The Pull Request
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Here are a few more tips that could be helpful throughout getting your PR merged:

3.7.4.1. Addressing reviews
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Now that you have opened a PR that has the necessary elements with it and got reviewers
assigned, you will hopefully begin to receive reviews soon from them. Please make sure
to address those review comments, change requests, etc to get your PR ready to be
merged.

.. note::
    When you addressed all of a reviewer's comments/requests, do not forget to
    re-request their next review!

3.7.4.2. GitHub Actions checks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

UXarray employs a number of GitHub Actions workflows (please refer to the `GitHub
Actions <https://docs.github.com/en/actions>`_ guide for detailed information) to make
sure our PRs, branches, etc. pass certain test scenarios such as the pre-commit hooks,
code test suite, and documentation generation. The pre-commit hooks workflow ensures
the code being proposed to be merged is complying with code standards. The test suite
workflows ensure that the changes are passing for those tests described in `3.7.3.1.

> Unit tests`_ for a matrix of various platforms and Python versions. The documentation

Reviewer 3: ```suggestion
Unit Tests`_ for a matrix of various platforms and Python versions. The documentation
```
If prior review comment is committed wrt section name

Author: That one wasn't committed/ Thanks though!

.. _contributing:

*******************
Contributor's Guide
*******************

Welcome to the Contributor's Guide for UXarray!

Welcome to the team! If you are reading this document, we hope that
you are already or soon-to-be a UXarray contributor, please keep reading!

1. Overview
===========

UXarray is a community-owned, open-development effort that is a result

> of the collaboration between `Project Raijin <https://raijin.ucar.edu/>`_

Reviewer 2: Say "NSF's Project Raijin" or remove "DOE's".

Author: Great catch! Added "NSF's ..."

and DOE’s `SEATS Project <https://seatstandards.org/>`_. Even though the
UXarray team has initiated and been expanding this package, outside
contributions are welcome and vital for its long-term sustainability.
Therefore, we invite other community members to become part of this
collaboration at any level of contribution.

> 1.1. Why the name "UXarray"?

Reviewer 2: This seems like it belongs in the README instead of the Contributors guide.  (Potential contributors should read the README)

Reviewer 2: We can assume they've already read the README if they're reading the Contributor's Guide.

Reviewer 2: Or just add that to the instructions.

Author: This makes sense. Moved this section to README, which also led me to adjust some headings as well as fix the use of the words "Uxarray"  in README.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^

There are many different ways to contribute to UXarray. Anyone can, for
example,

- Write or revise documentation (including this document)

- Implement data analysis operators for unstructured grids from scratch
  or from their workflows

- Develop example notebooks that demonstrate how a particular function
  is used

- Answer a support question

- Simply request a feature or report a bug

All of these activities are signicant contributions to the on-going
development and maintenance of UXarray.

1.3. About this guide
^^^^^^^^^^^^^^^^^^^^^

The UXarray content is hosted on GitHub, and through this document,
we aim to ease the community's experience with contributing to this project.
However, this guide might still be missing case-specific details; please do
not hesitate to reach out to us to consult any such cases.

.. note::
    Much of the information in this guide has been co-opted from the
    `GeoCAT <https://geocat.ucar.edu/pages/contributing.html>`_ project and
    `Project Pythia <https://projectpythia.org/contributing.html>`_.

1.4. Project Specific Resources
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Some important UXarray resources are as follows:

> - `Uxarray GitHub repository <https://github.com/UXARRAY/uxarray>`_ houses

Reviewer 3: ```suggestion
- `UXarray GitHub repository <https://github.com/UXARRAY/uxarray>`_ houses
```

  the open-source code base along with some significant documentation such
  as `Readme <https://github.com/UXARRAY/uxarray/blob/main/README.md>`_ and
  `UXarray draft API <https://github.com/UXARRAY/uxarray/blob/main/docs/user_api/uxarray_api.md>`_.

> - `Uxarray documentation <https://uxarray.readthedocs.io/en/latest/?badge=latest>`_

Reviewer 3: ```suggestion
- `UXarray documentation <https://uxarray.readthedocs.io/en/latest/?badge=latest>`_
```

another. As such, we will provide an overview of GitHub here and refer the
reader to other more comprehensive resources for detailed information about
it (such as GitHub's this
`Getting Started with Github <https://docs.github.com/en/get-started>`_
guide), but we will provide great details about how to configure a GitHub
development environment for UXarray development and how to contribute.

2.1. Getting started with GitHub and Git
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Contributing to UXarray requires using GitHub, as already mentioned, and
Git. The latter, Git, is an open source, command line tool for collaborative
software version control, while GitHub is an online, web-accessible service
that greatly simplifies using the powerful, yet often complex, Git.

.. note::
    GitHub operates entirely within a web browser. You do not need to
    install anything, but you will need to set up a free GitHub account. Git
    is a command line tool that most likely will need to be installed on
    your machine and run from a "terminal" window, AKA a "shell".

Using, and even just configuring, Git and GitHub are often the most

> daunting aspects of contributing to a GitHub hosted project. Here are

Reviewer 3: Should there be a hyperlink here to the steps, or will the steps be listed below this section (and/or added at a later time)?

Author: The next subsections are those steps. I added ":" at the end of the paragraph now to reflect this.
